src/plot.js

```diff
@@ -96,12 +95,38 @@ export function plot(options = {}) {
       .call(applyInlineStyles, style)
     .node();
 
-  for (const mark of marks) {
+  // A mark layout is responsible for:
+  // * applying the scales, possibly pushing a new one (for legends)
+  // * returning a filtered index and values to render, possibly new channels to rescale
+  const newScaleChannels = new Map();
+  const layouts = [];
+  const newOptions = {};
+  for (let i = 0; i < marks.length; ++i) {
+    const mark = marks[i];
+    const {channels} = layouts[i] = applyLayout.call(mark, mark.layout, [markIndex.get(mark)], scaleDescriptors, markChannels.get(mark), dimensions, options);
+    if (channels) {
+      for (const [, {scale, options: hint}] of channels) {
+        if (scale) newOptions[scale] = {...hint, ...options[scale]};
+      }
+      addScaleChannels(newScaleChannels, channels, newOptions);
+      markChannels.set(mark, channels);
+    }
+  }
+
+  // compute descriptors for new scales
+  const newScaleDescriptors = Scales(newScaleChannels, newOptions);
+  scaleDescriptors = {...newScaleDescriptors, ...scaleDescriptors};
+  scales = ScaleFunctions(scaleDescriptors);
+  for (const layout of layouts) {
+    if (layout.channels) layout.values = {...layout.values, ...applyScales(layout.channels, scales)};
```

Comment:
Similarly this is mutating *layout*, which was returned by *mark*.layout; we shouldn’t mutate an object we didn’t create (e.g. the mark might use memoization).